src/execution_impl.rs

```diff
@@ -13,88 +16,119 @@ use wasmer::{
 use wasmer_as::{Read as ASRead, StringPtr, Write as ASWrite};
 use wasmer_middlewares::Metering;
 
-/// `Print` ABI called by the webassembly VM
-fn print(env: &Env, arg: i32) {
-    let str_ptr = StringPtr::new(arg as u32);
-    println!(
-        "{}",
-        str_ptr
-            .read(env.wasm_env.memory.get_ref().expect("initialized memory"))
-            .unwrap()
-    );
+/// Utility function to call a WASM module (compiled from AssemblyScript) using
+/// high-level Rust types (by de/serializing a String).
+///
+/// The intend is to be able to expose any kind of function as an ABI:
+///
+/// ```ignore
+/// fn get_status() -> Result<NodeStatus> {
+///     Ok(typed_call(env, "0xAddressOfSCInLedger", "get_status", ())?)
+/// }
+/// ```
+pub fn typed_call<T: Serialize, R: DeserializeOwned>(
+    env: &Env,
+    address: &Address,
+    function: &str,
+    _args: T, // TODO: @adrien-zinger how do we pass function args in AssemblyScript?
+) -> Result<R> {
+    let value = call_module(env, address, function);
+    Ok(serde_json::from_str::<R>(&value.ret)?)
 }
 
 /// `Call` ABI called by the webassembly VM
+///
+/// Call an exported function in a WASM module at a given address
+///
 /// It take in argument the environment defined in env.rs
 /// this environment is automatically filled by the wasmer library
 /// And two pointers of string. (look at the readme in the wasm folder)
-fn call(env: &Env, address: i32, function: i32) -> i32 {
-    let memory = env.wasm_env.memory.get_ref().expect("initialized memory");
-    let addr_ptr = StringPtr::new(address as u32);
-    let func_ptr = StringPtr::new(function as u32);
-    let address = &addr_ptr.read(memory).unwrap();
-    let fnc = &func_ptr.read(memory).unwrap();
-    type GmSign = fn(&Address) -> Result<Bytecode>;
-    let get_module: GmSign = env.interface.get_module;
-    let module = &get_module(address).unwrap();
+pub fn call_module(env: &Env, address: &Address, function: &str) -> Response {
+    let get_module: fn(&Address) -> Result<Bytecode> = env.interface.get_module;
     sub_remaining_point(env, settings::METERING.call_price()).unwrap();
-    let value = exec(
+    let module = &get_module(address).unwrap();
+    exec(
         get_remaining_points(env),
         None,
         module,
-        fnc,
+        function,
         &[],
         &env.interface,
     )
-    .unwrap();
+    .unwrap() // TODO: Should return a Result<Response>
```

Comment:
Is there an issue for that ?

Reply:
Oups this todo isn't a real todo, the call function will always return an offset (i32)

Reply:
That's not true! This function isn't passed to WASM, it's `assembly_script_call_module` which is! So this function can take and return cool rust types :)

Reply:
May be yes. As I understand, the web assembly signature of the function is still
```wat
(import "massa" "assembly_script_call" (func $assembly/massa/call (param i32 i32) (result i32)))
```

So we can think that the rust object can be returned. Do you mean that you'll serialize to an AssemblyScript object the Result<Response>?

Reply:
never mind, I misunderstood :grimacing: